torchao/core/config.py

```diff
@@ -26,4 +33,214 @@ def _transform(
+        if hasattr(o, "_fields") and hasattr(
+            o, "_asdict"
+        ):  # Check for NamedTuple characteristics
+            asdict_data = o._asdict()
+            # Process each field to handle nested objects
+            processed_data = {k: self.encode_value(v) for k, v in asdict_data.items()}
+
+            return {
+                "_type": o.__class__.__name__,
+                "_version": getattr(o.__class__, "VERSION", 1),
+                "_data": processed_data,
+            }
+
+        # Handle dataclasses
+        if dataclasses.is_dataclass(o) and not isinstance(o, type):
+            data_dict = {}
+            # Process each field to handle nested objects
+            for f in dataclasses.fields(o):
+                if f.name != "VERSION":
+                    data_dict[f.name] = self.encode_value(getattr(o, f.name))
+
+            return {
+                # Only store the class name for dataclasses too
+                "_type": o.__class__.__name__,
+                "_version": getattr(o.__class__, "VERSION", 1),
+                "_data": data_dict,
+            }
+
+        # Handle torch.dtype
+        if hasattr(o, "__module__") and o.__module__ == "torch" and isinstance(o, type):
+            return {"_type": "torch.dtype", "_data": str(o).split(".")[-1]}
+
+        # Handle Layout objects
+        if hasattr(o, "__class__") and "Layout" in o.__class__.__name__:
+            return {
+                "_type": o.__class__.__name__,
+                "_data": {
+                    k: self.encode_value(v)
+                    for k, v in o.__dict__.items()
+                    if not k.startswith("_")
+                },
+            }
+
+        # Handle enum values
+        if isinstance(o, enum.Enum):
+            # Store the full path for enums to ensure uniqueness
+            return {"_type": f"{o.__class__.__name__}", "_data": o.name}
+
+        if isinstance(o, torch.dtype):
+            return {"_type": "torch.dtype", "_data": str(o).split(".")[-1]}
+
+        # For lists and dictionaries, recursively process their items
+        if isinstance(o, list):
+            return [self.encode_value(item) for item in o]
+
+        if isinstance(o, dict):
+            return {k: self.encode_value(v) for k, v in o.items()}
+
+        # Default case - let the parent class handle it
+        return super().default(o)
+
+    def encode_value(self, value):
+        """Helper method to recursively encode a value"""
+        # Try to use default for custom type
+        try:
+            # This will handle all our special cases and raise TypeError
+            # if it can't handle the type
+            result = self.default(value)
+            return result
+        except TypeError:
+            pass
+
+        # Default case - return as is
+        # (This will be processed by standard JSON encoder later)
+        return value
+
+
+def config_to_dict(config: AOBaseConfig) -> Dict[str, Any]:
```

Comment:
question: why not put this and `config_from_dict` as methods on `AOBaseConfig`?

Reply:
I think it's better to just expose `AOBaseConfig` rather than 2 separate functions. We need to expose `AOBaseConfig` anyway for users who wish to write custom configs, since we are removing the ability to pass in Callables into quantize_ in this release

Reply:
Sorry one other point is that for config_from_dict you dont know the base config youare trying to builld from and having this as the ABCs' class method feels kinda weird to me. These should mostly be implementation details for other libraries and really called by users